neo-codap/src/graph/graph.tsx

```diff
@@ -59,18 +95,119 @@ export class GraphComponent extends React.Component<IGraphProps, IGraphState> {
+            let nonNumericCount = 0,
+                nonEmptyCount = 0;
+            attr.values.forEach((value: string|number|undefined) => {
+                if ((value != null) && (value !== '')) {
+                    ++nonEmptyCount;
+                    if (!isFinite(Number(value))) {
+                        ++nonNumericCount;
+                        uniqueValues[value] = true;
+                    }
+                }
+            });
+
+            // attribute score is based on the ratio of non-numeric values to non-empty values
+            // with a boost based on the inverse of the number of unique values
+            // (so that attributes with fewer unique values score higher)
+            let score = nonEmptyCount > 0 ? nonNumericCount / nonEmptyCount : 0;
+            const numUniqueValues = Object.keys(uniqueValues).length;
+            score += numUniqueValues > 0 ? (1 / numUniqueValues) : 0;
+
+            return {id: attr.id, score, index};
+        });
+        if (scores.length > 0) {
+            scores.sort((a, b) => {
+                // in case of a tying score pick the left most one
+                if (a.score === b.score) {
+                    return a.index - b.index;
+                }
+                // otherwise, sort by descending score
+                return b.score - a.score;
+            });
+            const highestScore = scores[0];
+            return highestScore.score > 0 ? highestScore.id : undefined;
+        }
+        return undefined;
+    }
+
+    getLegendAttrType(srcDataSet: IDataSet, legendAttr: IAttribute) {
+
+        let nonEmptyCount = 0;
+        let min = 0;
+        let max = 0;
+        const numericValues: number[] = [];
+
+        legendAttr.values.forEach((value: IValueType) => {
+            if ((value !== undefined) && (value !== null) && (value !== '')) {
+                nonEmptyCount++;
+                const numericValue = Number(value);
+                if (isFinite(numericValue)) {
+                    if (numericValues.length === 0) {
+                        min = max = numericValue;
+                    }
+                    else {
+                        min = Math.min(min, numericValue);
+                        max = Math.max(max, numericValue);
+                    }
+                    numericValues.push(numericValue);
+                }
+            }
+        });
+
+        const isNumeric = (nonEmptyCount > 0) &&
+                          (numericValues.length > 0) &&
+                          ((numericValues.length / nonEmptyCount) >= 0.5);
+        const quantiles: ILegendQuantile[] = [];
+        const values: IValueType[] = [];
+
+        if (isNumeric) {
+            numericValues.sort((a, b) => a - b);
+            const numQuantiles = Math.min(5, numericValues.length);
+            const lastIndex = numericValues.length - 1;
+
+            // adapted from CODAP's DG.MathUtilities.nQuantileValues()
```

Comment:
If this codebase continues, there may be other use for such a function, so eventually it may make sense to put it in a separately callable utility function.